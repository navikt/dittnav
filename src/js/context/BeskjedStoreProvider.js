import React, { createContext, useReducer } from 'react';
import { arrayOf, node } from 'prop-types';
import Api from '../Api';
import {
  ADD_BESKJEDER,
  ADD_INAKTIVE_BESKJEDER,
  ADD_INAKTIV_BESKJED, REMOVE_BESKJED,
} from '../types/Actions';
import BeskjedType from '../types/BeskjedType';

const initialState = (beskjeder, inaktiveBeskjeder) => ({
  beskjeder,
  inaktiveBeskjeder,
});

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_BESKJEDER:
      return {
        ...state,
        beskjeder: action.payload,
      };
    case ADD_INAKTIVE_BESKJEDER:
      return {
        ...state,
        inaktiveBeskjeder: action.payload,
      };
    case REMOVE_BESKJED:
      Api.postDone({
        eventId: action.payload.eventId,
        uid: action.payload.uid,
      });
      return {
        ...state,
        beskjeder: state.beskjeder.filter(b => action.payload.uid !== b.uid),
      };
    case ADD_INAKTIV_BESKJED:
      return {
        ...state,
        inaktiveBeskjeder: [...state.inaktiveBeskjeder, action.payload],
      };
    default:
      return state;
  }
};

export const BeskjedStoreContext = createContext(null);

const BeskjedStoreProvider = ({ beskjeder, inaktiveBeskjeder, children }) => {
  const [state, dispatch] = useReducer(reducer, initialState(beskjeder, inaktiveBeskjeder));
  const value = { state, dispatch };

  return (
    <BeskjedStoreContext.Provider value={value}>
      {children}
    </BeskjedStoreContext.Provider>
  );
};

BeskjedStoreProvider.propTypes = {
  beskjeder: arrayOf(BeskjedType),
  inaktiveBeskjeder: arrayOf(BeskjedType),
  children: node.isRequired,
};

BeskjedStoreProvider.defaultProps = {
  beskjeder: null,
  inaktiveBeskjeder: null,
};

export default BeskjedStoreProvider;
