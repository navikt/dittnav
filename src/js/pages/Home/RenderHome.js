import React, { useEffect, useState } from 'react';
import useBeskjedStore from '../../hooks/useBeskjedStore';
import PageFrame from '../PageFrame';
import Home from './Home';
import { ADD_BESKJEDER } from '../../types/Actions';
import ApiType from '../../types/ApiType';

const RenderHome = ({ api }) => {
  const [data, setData] = useState({
    oppfolging: null,
    meldekort: null,
    person: null,
    identifikator: null,
    paabegynteSoknader: null,
    mininnboks: [],
    sakstema: { antallSakstema: 0, sakstemaList: [] },
    innlogging: null,
    oppgaver: null,
    innbokser: null,
    antallBrukernotifikasjoner: 0,
    errors: [],
    fetching: 0,
    oppfolgingHasLoaded: false,
  });

  const { dispatch } = useBeskjedStore();

  const handleOppfolgingError = () => {
    setData(d => ({ ...d, errors: [...d.errors, 'error.baksystemer'], fetching: d.fetching + 1, oppfolgingHasLoaded: true }));
  };

  const incrementFetching = () => {
    setData(d => ({ ...d, fetching: d.fetching + 1 }));
  };

  useEffect(
    () => {
      const handleError = () => {
        incrementFetching();
        setData(d => ({ ...d, errors: [...d.errors, 'error.baksystemer'] }));
      };

      api.fetchBeskjeder()
        .then((r) => {
          incrementFetching();
          dispatch({ type: ADD_BESKJEDER, payload: r });
        }).catch(handleError);

      api.fetchOppgaver()
        .then((r) => {
          setData(d => ({ ...d, oppgaver: r, fetching: d.fetching + 1 }));
        }).catch(handleError);

      api.fetchInnbokser()
        .then((r) => {
          setData(d => ({ ...d, innbokser: r, fetching: d.fetching + 1 }));
        }).catch(handleError);

      /* Skrur midlertidig av denne for å se om det har en effekt på cpu-bruk i database.
      api.fetchAntallBrukernotifikasjoner()
        .then((r) => {
          setData(d => ({ ...d, antallBrukernotifikasjoner: r, fetching: d.fetching + 1 }));
        }).catch(handleError);
       */

      api.fetchInnlogging()
        .then((r) => {
          setData(d => ({ ...d, innlogging: r, fetching: d.fetching + 1 }));
        }).catch(handleError);

      api.fetchOppfolging()
        .then((r) => {
          setData(d => ({ ...d, oppfolging: r, oppfolgingHasLoaded: true, fetching: d.fetching + 1 }));
        }).catch(handleOppfolgingError);

      api.fetchMeldekort()
        .then((r) => {
          setData(d => ({ ...d, meldekort: r, fetching: d.fetching + 1 }));
        }).catch(handleError);

      api.fetchPersonNavn()
        .then((r) => {
          setData(d => ({ ...d, person: r, fetching: d.fetching + 1 }));
        }).catch(() => {
          api.fetchPersonIdent()
            .then(r => {
              setData(d => ({ ...d, identifikator: r, errors: [...d.errors, 'error.baksystemer'], fetching: d.fetching + 1 }));
            })
            .catch(handleError);
        });

      api.fetchSaker()
        .then((r) => {
          const { feilendeBaksystem } = r;
          if (feilendeBaksystem.length > 0) {
            setData(d => ({ ...d, paabegynteSoknader: r, fetching: d.fetching + 1, errors: [...d.errors, 'error.baksystemer'] }));
          } else {
            setData(d => ({ ...d, paabegynteSoknader: r, fetching: d.fetching + 1 }));
          }
        }).catch(handleError);

      api.fetchMeldinger()
        .then((r) => {
          setData(d => ({ ...d, mininnboks: r, fetching: d.fetching + 1 }));
        }).catch(handleError);

      api.fetchSakstema()
        .then((r) => {
          setData(d => ({ ...d, sakstema: r, fetching: d.fetching + 1 }));
        }).catch(handleError);
    }, [],
  );

  const uniqueErrors = data.errors.filter((item, i, ar) => ar.indexOf(item) === i);
  const loading = data.fetching < 10;

  return (
    <PageFrame uniqueErrors={uniqueErrors}>
      <Home data={data} loading={loading} />
    </PageFrame>
  );
};

RenderHome.propTypes = {
  api: ApiType.isRequired,
};

export default RenderHome;
