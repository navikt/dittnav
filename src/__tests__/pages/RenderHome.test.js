import React from 'react';
import ReactDOM from 'react-dom';
import RenderHome from 'js/pages/Home/RenderHome';
import ShallowRenderer from 'react-test-renderer/shallow';
import wrapIntl from 'js/IntlTestHelper';
import BeskjedStoreProvider from '../../js/context/BeskjedStoreProvider';
const ReactTestRenderer = require('react-test-renderer');

/* eslint-disable no-unused-vars */
const mockApi = () => (
  {
    fetchBeskjeder: () => new Promise((resolve, reject) => {}),
    fetchOppgaver: () => new Promise((resolve, reject) => {}),
    fetchInnbokser: () => new Promise((resolve, reject) => {}),
    fetchInaktiveBeskjeder: () => new Promise((resolve, reject) => {}),
    fetchInaktiveOppgaver: () => new Promise((resolve, reject) => {}),
    fetchInaktiveInnbokser: () => new Promise((resolve, reject) => {}),
    fetchInnlogging: () => new Promise((resolve, reject) => {}),
    fetchOppfolging: () => new Promise((resolve, reject) => {}),
    fetchMeldekort: () => new Promise((resolve, reject) => {}),
    fetchPersonNavn: () => new Promise((resolve, reject) => {}),
    fetchPersonIdent: () => new Promise((resolve, reject) => {}),
    fetchSaker: () => new Promise((resolve, reject) => {}),
    fetchMeldinger: () => new Promise((resolve, reject) => {}),
    fetchSakstema: () => new Promise((resolve, reject) => {}),
    fetchHendelser: () => new Promise((resolve, reject) => {}),
  }
);
/* eslint-enable no-unused-vars */

const flushPromises = () => (
  new Promise(resolve => setImmediate(resolve))
);

it('renders without crashing', () => {
  const div = document.createElement('div');

  ReactDOM.render(wrapIntl(
    <BeskjedStoreProvider>
      <RenderHome api={mockApi()} />
    </BeskjedStoreProvider>,
  ), div);
  ReactDOM.unmountComponentAtNode(div);
});

it('expect Login page rendering', () => {
  const api = mockApi();
  const component = ReactTestRenderer.create(wrapIntl(
    <BeskjedStoreProvider>
      <RenderHome api={api} />
    </BeskjedStoreProvider>,
  ));
  expect(component.toJSON()).toMatchSnapshot();
});

it('expect Oppfolging fetching', async () => {
  const api = mockApi();
  api.fetchOppfolging = () => new Promise((resolve, reject) => { // eslint-disable-line no-unused-vars
    resolve(
      {
        erBrukerUnderOppfolging: false,
      },
    );
  });
  const renderer = new ShallowRenderer();
  renderer.render(wrapIntl(<RenderHome api={api} />));
  const component = renderer.getRenderOutput();
  await flushPromises();

  expect(component).toMatchSnapshot();
});

it('expect MeldekortInfo fetching', async () => {
  const api = mockApi();
  api.fetchMeldekort = () => new Promise((resolve, reject) => { // eslint-disable-line no-unused-vars
    resolve(
      {
        etterregistrerteMeldekort: 0,
        meldekortbruker: true,
        nyeMeldekort: {
          antallNyeMeldekort: 3,
          nesteInnsendingAvMeldekort: null,
          nesteMeldekort: {
            fra: '2019-09-09',
            kanSendesFra: '2019-09-21',
            risikererTrekk: true,
            sisteDatoForTrekk: '2019-09-30',
            til: '2019-09-22',
            uke: '37-38',
          },
        },
        resterendeFeriedager: 0,
      },
    );
  });

  const renderer = new ShallowRenderer();
  renderer.render(wrapIntl(<RenderHome api={api} />));
  const component = renderer.getRenderOutput();
  await flushPromises();

  expect(component).toMatchSnapshot();
});

it('expect Personalia navn fetching', async () => {
  const api = mockApi();
  api.fetchPersonNavn = () => new Promise((resolve, reject) => { // eslint-disable-line no-unused-vars
    resolve(
      {
        navn: 'VINAYAGUM-MASK AMIZIC',
      },
    );
  });
  const renderer = new ShallowRenderer();
  renderer.render(wrapIntl(<RenderHome api={api} />));
  const component = renderer.getRenderOutput();
  await flushPromises();

  expect(component).toMatchSnapshot();
});

it('expect Personalia ident fetching', async () => {
  const api = mockApi();
  api.fetchPersonIdent = () => new Promise((resolve, reject) => { // eslint-disable-line no-unused-vars
    resolve(
      {
        ident: 123,
      },
    );
  });
  const renderer = new ShallowRenderer();
  renderer.render(wrapIntl(<RenderHome api={api} />));
  const component = renderer.getRenderOutput();
  await flushPromises();

  expect(component).toMatchSnapshot();
});


it('expect PaabegynteSoknader fetching', async () => {
  const api = mockApi();

  api.fetchSaker = () => new Promise((resolve, reject) => { // eslint-disable-line no-unused-vars
    resolve(
      {
        url: 'https://tjenester-t6.nav.no/',
        antallPaabegynte: 2,
        feilendeBaksystem: ['hello'],
      },
    );
  });

  const renderer = new ShallowRenderer();
  renderer.render(wrapIntl(<RenderHome api={api} />));
  const component = renderer.getRenderOutput();
  await flushPromises();

  expect(component).toMatchSnapshot();
});
