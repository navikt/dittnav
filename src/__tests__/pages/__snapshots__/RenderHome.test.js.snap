// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect Login page rendering 1`] = `
<main
  role="main"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="maincontent side-innhold"
      >
        <div
          className="col-md-12"
          id="dittnav-main-container"
        >
          <section
            className="infomeldinger-list"
          >
            <h1
              className="skjermleser"
            >
              <span>
                Varsler
              </span>
            </h1>
            <a
              className="lenkepanel infomelding beskjed lenkepanel--border"
              href="https://www.nav.no/no/person/innhold-til-person-forside/nyttig-a-vite/koronavirus--informasjon-fra-nav"
              onClick={null}
            >
              <div
                className="lenkepanel__innhold"
              >
                <div
                  className="lenkepanel__ikon"
                >
                  <svg
                    height="48px"
                    version="1.1"
                    viewBox="0 0 48 48"
                    width="48px"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlnsXlink="http://www.w3.org/1999/xlink"
                  >
                    <g
                      fill="none"
                      fillRule="evenodd"
                      stroke="none"
                      strokeWidth="1"
                    >
                      <g
                        transform="translate(-382.000000, -303.000000)"
                      >
                        <g
                          transform="translate(382.000000, 303.000000)"
                        >
                          <circle
                            cx="24"
                            cy="24"
                            fill="#C2EAF7"
                            r="24"
                          />
                          <path
                            d="M23.5,12 C29.8410001,12 35,17.159 35,23.5 C35,29.8410001 29.8410001,35 23.5,35 C17.159,35 12,29.8410001 12,23.5 C12,17.159 17.159,12 23.5,12 Z M23.5,13 C17.71,13 13,17.71 13,23.5 C13,29.2900001 17.71,34 23.5,34 C29.2900001,34 34,29.2900001 34,23.5 C34,17.71 29.2900001,13 23.5,13 Z M21.5,21 L23.5,21 C23.7453334,21 23.9495803,21.1769877 23.9919397,21.4101619 L24,21.5 L24,30 L26.5,30 C26.7760001,30 27,30.2240001 27,30.5 C27,30.7453334 26.8230124,30.9495803 26.5898382,30.9919397 L26.5,31 L20.5,31 C20.224,31 20,30.7760001 20,30.5 C20,30.2546667 20.1769877,30.0504198 20.4101619,30.0080604 L20.5,30 L23,30 L23,22 L21.5,22 C21.224,22 21,21.776 21,21.5 C21,21.2546667 21.1769877,21.0504198 21.4101619,21.0080604 L21.5,21 L23.5,21 Z M23,16.5 C23.5522848,16.5 24,16.9477153 24,17.5 C24,18.0522848 23.5522848,18.5 23,18.5 C22.4477153,18.5 22,18.0522848 22,17.5 C22,16.9477153 22.4477153,16.5 23,16.5 Z"
                            fill="#3E3832"
                            id="Combined-Shape"
                          />
                        </g>
                      </g>
                    </g>
                  </svg>
                </div>
                <div>
                  <h2
                    className="typo-undertittel lenkepanel__heading"
                  >
                    <span>
                      Koronavirus – informasjon fra NAV
                    </span>
                  </h2>
                  <p
                    className="typo-normal"
                  >
                    <span>
                      Om sykmelding, omsorgspenger, permittering, papirmeldekort, arbeidsmarkedstiltak, akuttolk for døve m.fl.
                    </span>
                  </p>
                </div>
              </div>
              <span
                className="lenkepanel__indikator"
              />
            </a>
          </section>
          <div
            className="dittnav-lenkepanel-top-container"
          >
            <div
              className="dittnav-lenkepanel-liten"
              id={null}
            >
              <a
                alt="Dine saker"
                className="lenkepanel dittnav-lenkepanel-liten-item lenkepanel--border"
                href="http://localhost:9222/saksoversikt"
              >
                <span
                  className="typo-undertittel lenkepanel__heading"
                >
                  <span>
                    Dine saker
                  </span>
                </span>
                <span
                  className="lenkepanel__indikator"
                />
              </a>
              <a
                alt="Utbetalinger"
                className="lenkepanel dittnav-lenkepanel-liten-item lenkepanel--border"
                href="http://localhost:9222/utbetalingsoversikt"
              >
                <span
                  className="typo-undertittel lenkepanel__heading"
                >
                  <span>
                    Dine utbetalinger
                  </span>
                </span>
                <span
                  className="lenkepanel__indikator"
                />
              </a>
              <a
                alt="Innboks"
                className="lenkepanel dittnav-lenkepanel-liten-item last lenkepanel--border"
                href="http://localhost:9222/mininnboks"
              >
                <span
                  className="typo-undertittel lenkepanel__heading"
                >
                  <span>
                    Innboks
                  </span>
                </span>
                <span
                  className="lenkepanel__indikator"
                />
              </a>
            </div>
          </div>
          <h2
            className="typo-undertittel flere-tjenester__subheader"
          >
            <span>
              Flere tjenester
            </span>
          </h2>
          <div
            className="flere-tjenester"
          >
            <nav
              className="flere-tjenester__links"
            >
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="https://arbeidsplassen.nav.no/stillinger"
                  >
                    Ledige stillinger
                  </a>
                </p>
              </div>
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="http://localhost:9222/pselv/publisering/uforetrygd.jsf?context=ut"
                  >
                    Uføretrygd
                  </a>
                </p>
              </div>
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="https://foreldrepenger.nav.no"
                  >
                    Dine foreldrepenger
                  </a>
                </p>
              </div>
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="http://localhost:9222/aktivitetsplan"
                  >
                    Aktivitetsplan
                  </a>
                </p>
              </div>
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="http://localhost:9222/meldekort/om-meldekort"
                  >
                    Meldekort
                  </a>
                </p>
              </div>
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="http://localhost:9222/veiledearbeidssoker/mistet-jobben/registrering"
                  >
                    Registrer deg som arbeidssøker
                  </a>
                </p>
              </div>
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="https://stillingsok.nav.no/pam-stillingsok/lagrede-sok"
                  >
                    Dine stillingssøk
                  </a>
                </p>
              </div>
              <div
                className="flere-tjenester__link-container"
              >
                <p
                  className="typo-normal"
                >
                  <a
                    className="lenke flere-tjenester__link"
                    data-ga="Dittnav/Lenkeliste"
                    href="http://localhost:9222/person/personopplysninger"
                  >
                    Personopplysninger
                  </a>
                </p>
              </div>
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;

exports[`expect MeldekortInfo fetching 1`] = `
<RenderHome
  api={
    Object {
      "fetchHendelser": [Function],
      "fetchMeldekort": [Function],
      "fetchMeldinger": [Function],
      "fetchOppfolging": [Function],
      "fetchPersonIdent": [Function],
      "fetchPersonNavn": [Function],
      "fetchSaker": [Function],
      "fetchSakstema": [Function],
    }
  }
/>
`;

exports[`expect Oppfolging fetching 1`] = `
<RenderHome
  api={
    Object {
      "fetchHendelser": [Function],
      "fetchMeldekort": [Function],
      "fetchMeldinger": [Function],
      "fetchOppfolging": [Function],
      "fetchPersonIdent": [Function],
      "fetchPersonNavn": [Function],
      "fetchSaker": [Function],
      "fetchSakstema": [Function],
    }
  }
/>
`;

exports[`expect PaabegynteSoknader fetching 1`] = `
<RenderHome
  api={
    Object {
      "fetchHendelser": [Function],
      "fetchMeldekort": [Function],
      "fetchMeldinger": [Function],
      "fetchOppfolging": [Function],
      "fetchPersonIdent": [Function],
      "fetchPersonNavn": [Function],
      "fetchSaker": [Function],
      "fetchSakstema": [Function],
    }
  }
/>
`;

exports[`expect Personalia ident fetching 1`] = `
<RenderHome
  api={
    Object {
      "fetchHendelser": [Function],
      "fetchMeldekort": [Function],
      "fetchMeldinger": [Function],
      "fetchOppfolging": [Function],
      "fetchPersonIdent": [Function],
      "fetchPersonNavn": [Function],
      "fetchSaker": [Function],
      "fetchSakstema": [Function],
    }
  }
/>
`;

exports[`expect Personalia navn fetching 1`] = `
<RenderHome
  api={
    Object {
      "fetchHendelser": [Function],
      "fetchMeldekort": [Function],
      "fetchMeldinger": [Function],
      "fetchOppfolging": [Function],
      "fetchPersonIdent": [Function],
      "fetchPersonNavn": [Function],
      "fetchSaker": [Function],
      "fetchSakstema": [Function],
    }
  }
/>
`;
